Feature ss 4

patrick29 opened the pull request:
**Description:**

Created the initial structure for Job
Created endpoints to save/update/find/delete the Job entity
Created table through the Flyway for Job entity

**How to test:**

Call the endpoints and validate that all is working as expected. 
Try to use the example on Postman: [link](https://www.getpostman.com/collections/67d7406787323b00ff72)

**Release plan:**

anytime

patrick29 committed:
Creating first version JobEntity
Creating endpoints save, find, update, delete
Creating DTO classes for JobEntity
Creating test cases for JobEntity's DTO
Creating test cases for JobEntityRepository

patrick29 committed:
Merge remote-tracking branch 'origin/feature-SS-13' into feature-SS-4

patrick29 committed:
Refactoring method name
Cleaning code

patrick29 committed:
Merge branch 'staging' into feature-SS-4

# Conflicts:
#	src/main/java/com/shigotosagashi/jobdomainservice/controller/JobController.java

patrick29 committed:
Merge with branch staging

edylle commented on src/main/java/com/shigotosagashi/jobdomainservice/controller/JobController.java:
```diff
@@ -20,30 +21,31 @@
-      @Valid @RequestBody com.shigotosagashi.jobdomainservice.dto.in.ExampleEntityDto example) {
-    return ExampleEntityDto.fromEntity(exampleEntityRepository.save(example.toEntity()));
+  public JobOut save(
+      @Valid @RequestBody JobIn example) {
```
Rename `example` to something that makes more sense, like `job`

edylle commented on src/main/java/com/shigotosagashi/jobdomainservice/controller/JobController.java:
```diff
@@ -20,30 +21,31 @@
-  public ExampleEntityDto save(
-      @Valid @RequestBody com.shigotosagashi.jobdomainservice.dto.in.ExampleEntityDto example) {
-    return ExampleEntityDto.fromEntity(exampleEntityRepository.save(example.toEntity()));
+  public JobOut save(
```
NABD, but you could've saved a couple of lines in these methods. There's no need to write the parameters in new lines.

```
public JobOut save(@Valid @RequestBody JobIn example)
public JobOut update(@PathVariable String id, @Valid @RequestBody JobIn example)...
```

edylle commented on src/main/java/com/shigotosagashi/jobdomainservice/dto/out/JobOut.java:
```diff
@@ -0,0 +1,34 @@
+  public final long creationDate;
+  public final long expirationDate;
+
+  public JobOut(String id, String title, String description, long creationDate, long expirationDate) {
```
This line is too long. Please, write each parameter in a new line..

```
  public JobOut(String id,
                String title,
                String description,
                long creationDate,
                long expirationDate) {
```

edylle commented on src/main/java/com/shigotosagashi/jobdomainservice/dto/in/JobIn.java:
```diff
@@ -0,0 +1,49 @@
+  public final String description;
+  public final Optional<Long> creationDate;
+  @NotNull(message = "Missing required field expirationDate")
+  public final long expirationDate;
```
For DTOs, prefer using wrappers. Use Long

patrick29 replied:
Corrected

edylle commented on src/main/java/com/shigotosagashi/jobdomainservice/dto/out/JobOut.java:
```diff
@@ -0,0 +1,34 @@
+  public final String title;
+  public final String description;
+  public final long creationDate;
+  public final long expirationDate;
```
Wrappers

patrick29 replied:
Corrected

edylle commented on src/main/java/com/shigotosagashi/jobdomainservice/domain/JobEntity.java:
```diff
@@ -0,0 +1,54 @@
+        .title(UUID.randomUUID().toString())
+        .description(UUID.randomUUID().toString())
+        .creationDate(Instant.now().toEpochMilli())
+        .expirationDate(Instant.now().toEpochMilli() + new Random().nextLong())
```
Adding a random number to epoch milli could lead us to weird and hard to test dates. There's no need for this.
If you'd want to validate that the expiration date is after the creation one, you should do that inside the test itself. For example:

```
    Instant currentDate = Instant.now();
    Instant expirationDate = currentDate.plus(new Random().nextInt(100) + 1, ChronoUnit.MINUTES);

    assertThat(expirationDate).isAfter(currentDate);
```

edylle commented on src/main/java/com/shigotosagashi/jobdomainservice/dto/in/JobIn.java:
```diff
@@ -0,0 +1,49 @@
+        UUID.randomUUID().toString(),
+        UUID.randomUUID().toString(),
+        Optional.of(Instant.now().toEpochMilli()),
+        Instant.now().toEpochMilli() + new Random().nextLong()
```
Same as in JobEntity. No need for a random long

edylle commented on src/main/java/com/shigotosagashi/jobdomainservice/repository/JobEntityRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+  List<JobEntity> findByTitleLike(String title);
+  List<JobEntity> findByDescriptionLike(String description);
+  List<JobEntity> findByExpirationDateAfter(long datetime);
```
We're not actually using any of these methods. I would remove them.
And when we go to real production, we won't execute "like" queries in database anyway.

edylle commented on src/main/resources/db/migration/V002__job.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE JOB_ENTITY (
```
You deleted ExampleEntity related code. So in this migration file, you should drop `EXAMPLE_ENTITY` table.

patrick29 replied:
OK. Thinking now, the first migration file shouldn't be deleted? Because the `EXAMPLE_ENTITY` table will never go to production environment.

edylle replied:
In a real scenario, we must not delete migration files. Flyt would try to validate the version `001` against database, but this version wouldn't exist in our code.
As this project here is in its inicial phase and it's not on production yet, so it's up to you: either you delete the `V001` migration file and then you rename this one to be the new `V001`, or you add the `DROP TABLE` here (in version 002).

edylle commented on src/test/java/com/shigotosagashi/jobdomainservice/dto/in/JobInTest.java:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  @Test
+  void toEntity_whenIdIsIformed_transformsCorrectly() {
```
Typo: "Informed"

patrick29 replied:
Corrected

edylle commented on src/test/java/com/shigotosagashi/jobdomainservice/dto/out/JobOutTest.java:
```diff
@@ -0,0 +1,21 @@
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+
```
🙃

patrick29 replied:
What is that ?

edylle commented on src/test/java/com/shigotosagashi/jobdomainservice/repository/JobEntityRepositoryTest.java:
```diff
@@ -0,0 +1,97 @@
+  @Autowired
+  private JobEntityRepository repository;
+
+
```
🙃

patrick29 replied:
?

patrick29 replied:
Is the extra line?

edylle replied:
Yes, that's the extra line. Sorry if I was not clear enough.

edylle commented on src/test/java/com/shigotosagashi/jobdomainservice/repository/JobEntityRepositoryTest.java:
```diff
@@ -0,0 +1,97 @@
+
+
+  @Test
+  void findByTitleLike_findsExpectedEntity() {
```
You're not testing the "like" query in these methods.

patrick29 committed:
Renaming parameters
Cleaning unnecessary lines
Changing the variables types on Wrappers
Changing the queries to search from: 'Like' to: 'Containg'
Creating test cases to test properly queries
Editing script to drop EXAMPLE_ENTITY table

patrick29 committed:
Removing the random value for expirationDate
Removing unused methods from JobEntityRepository
